src/util/segmentation/operations/correction.js

```diff
@@ -0,0 +1,524 @@
+  });
+}
+
+/**
+ * SnapPointsToGrid - Snap the freehand points to the labelmap grid and attach a label for each node.
+ *
+ * @param  {Object[]} points An array of points drawn by the user.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {Object} evt The cornerstone event.
+ * @returns {Object[]}
+ */
+function snapPointsToGrid(evt) {
+  const { operationData } = evt;
+  const { pixelData, points } = operationData;
+
+  const { image } = evt.detail;
+  const cols = image.width;
+  const rows = image.height;
+
+  const nodes = [];
+
+  for (let i = 0; i < points.length; i++) {
+    const point = points[i];
+
+    let x = Math.floor(point.x);
+    let y = Math.floor(point.y);
+
+    // Clamp within the confines of the image.
+    x = clip(x, 0, cols - 1);
+    y = clip(y, 0, rows - 1);
+
+    const lastNode = nodes[nodes.length - 1];
+
+    // Skip double counting of closely drawn freehand points.
+    if (lastNode && x === lastNode.x && y === lastNode.y) {
+      continue;
+    }
+
+    nodes.push({
+      x,
+      y,
+      segment: pixelData[y * cols + x],
+    });
+  }
+
+  return nodes;
+}
+/**
+ * SimpleScissorOperation - Check if the operation is a simple scissors
+ * add/remove, and performs it if so.
+ * @param  {Object[]} nodes - The nodes snapped to the grid.
+ * @param  {Object} points - An array of points drawn by the user.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {Object} evt The cornerstone event.
+ * @param  {number} segmentIndex
+ * @returns {boolean} Returns true if the operation was simple.
+ */
+function simpleScissorOperation(nodes, evt, segmentIndex) {
+  let allInside = true;
+  let allOutside = true;
+
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+
+    if (node.segment === segmentIndex) {
+      allOutside = false;
+    } else {
+      allInside = false;
+    }
+
+    if (!allInside && !allOutside) {
+      break;
+    }
+  }
+
+  if (allOutside) {
+    logger.warn('The line never intersects a segment.');
+    fillInside(evt);
+
+    return true;
+  } else if (allInside) {
+    logger.warn('The line is only ever inside the segment.');
+    eraseInside(evt);
+
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * performOperation - Performs the given add/subtract operation using a modification of the Tobias Heimann Correction Algorithm:
+ * The algorithm is described in full length in Tobias Heimann's diploma thesis (MBI Technical Report 145, p. 37 - 40).
+ *
+ * @param  {Object} operation The operation.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {UInt16Array} workingLabelMap A copy of the labelmap for processing purposes.
+ * @param  {number} segmentIndex The label of the tool being used.
+ * @param  {Object} evt The cornerstone event.
+ */
+function performOperation(
+  operation,
+  pixelData,
+  workingLabelMap,
+  segmentIndex,
+  evt
+) {
+  const eventData = evt.detail;
+  const { image } = eventData;
```

Comment:
Do you use image or eventData?

```js
const { width: cols, height: rows } = evt.detail.image;
```

Reply:
This is a cool destructuring, felt like I should've been able to do something like that, thanks!